controller/handler_ctrl/accept.go

```diff
@@ -59,64 +59,75 @@ func (self *CtrlAccepter) Bind(binding channel.Binding) error {
+		if versionValue, found := ch.Underlay().Headers()[channel.HelloVersionHeader]; found {
+			if versionInfo, err := self.network.VersionProvider.EncoderDecoder().Decode(versionValue); err == nil {
+				r.VersionInfo = versionInfo
+				log = log.WithField("version", r.VersionInfo.Version).
+					WithField("revision", r.VersionInfo.Revision).
+					WithField("buildDate", r.VersionInfo.BuildDate).
+					WithField("os", r.VersionInfo.OS).
+					WithField("arch", r.VersionInfo.Arch)
 			} else {
-				log.Warn("no advertised listeners")
-			}
-			if val, found := ch.Underlay().Headers()[int32(ctrl_pb.ContentType_RouterMetadataHeader)]; found {
-				log.Debug("router reported listeners using listeners header")
-				routerMetadata := &ctrl_pb.RouterMetadata{}
-				if err = proto.Unmarshal(val, routerMetadata); err != nil {
-					log.WithError(err).Error("unable to unmarshall router metadata value")
-				}
-				r.SetMetadata(routerMetadata)
+				return errors.Wrap(err, "could not parse version info from router hello, closing router connection")
 			}
 		} else {
 			return errors.New("no version info header, closing router connection")
 		}
 
-		r.Control = ch
-		r.ConnectTime = time.Now()
-		if err := binding.Bind(newBindHandler(self.heartbeatOptions, r, self.network, self.xctrls)); err != nil {
-			return errors.Wrap(err, "error binding router")
+		r.Listeners = nil
+		if val, found := ch.Underlay().Headers()[int32(ctrl_pb.ContentType_ListenersHeader)]; found {
+			listeners := &ctrl_pb.Listeners{}
+			if err = proto.Unmarshal(val, listeners); err != nil {
+				log.WithError(err).Error("unable to unmarshall listeners value")
+			} else {
+				r.SetLinkListeners(listeners.Listeners)
+				for _, listener := range listeners.Listeners {
+					log.WithField("address", listener.GetAddress()).
+						WithField("protocol", listener.GetProtocol()).
+						WithField("costTags", listener.GetCostTags()).
+						Debug("router listener")
+				}
+			}
+		} else {
+			log.Info("no advertised listeners")
 		}
 
-		if self.traceHandler != nil {
-			binding.AddPeekHandler(self.traceHandler)
+		if val, found := ch.Underlay().Headers()[int32(ctrl_pb.ContentType_RouterMetadataHeader)]; found {
+			routerMetadata := &ctrl_pb.RouterMetadata{}
+			if err = proto.Unmarshal(val, routerMetadata); err != nil {
+				log.WithError(err).Error("unable to unmarshall router metadata value")
+			}
+			r.SetMetadata(routerMetadata)
 		}
+	} else {
+		return errors.New("channel provided no headers, closing router connection as version info not provided")
+	}
 
-		log.Infof("accepted new router connection [r/%s]", r.Id)
+	r.Control = ch
+	r.ConnectTime = time.Now()
+	if err := binding.Bind(newBindHandler(self.heartbeatOptions, r, self.network, self.xctrls)); err != nil {
+		return errors.Wrap(err, "error binding router")
+	}
 
-		self.network.ConnectRouter(r)
+	if self.traceHandler != nil {
+		binding.AddPeekHandler(self.traceHandler)
 	}
+
+	log.Info("accepted new router connection")
```

Comment:
why lose the existing, presumably useful, `r.Id`? is the `[r/Id]` not useful?

Reply:
There's already a logger field for routerId added at the top, so this was redundant.